theatre/studio/src/uiComponents/usePresence.tsx

```diff
@@ -0,0 +1,231 @@
+      StrictRecord<
+        StudioSheetItemKey,
+        StrictRecord<string, HoveredWithImportance>
+      >
+    >,
+  )
+
+  let lastRelationId = 0
+
+  return {
+    addRelationships(itemKey, relationships) {
+      const relId = String(++lastRelationId)
+      const undoAtPaths = relationships.map((rel) => {
+        const hovered: HoveredWithImportance = {
+          importance: rel.importance ?? 0,
+          relationship: rel.relationship,
+        }
+        const path = [rel.affects, itemKey, relId]
+        relations.setIn(path, hovered)
+        return path
+      })
+      return () => {
+        for (const pathToUndo of undoAtPaths) {
+          relations.setIn(pathToUndo, undefined)
+        }
+      }
+    },
+    useFocused(itemKey) {
+      const focusD = useMemo(() => {
+        if (!itemKey) return
+        const currentD = currentPresence.derivation
+        const primaryFocusDer = valueDerivation(atom.pointer[itemKey])
+        const relationsDer = valueDerivation(relations.pointer[itemKey])
+        return prism(() => {
+          const primary = primaryFocusDer.getValue()
+          if (primary) {
+            return Hovered.Hovered
+          } else {
+            const related = relationsDer.getValue()
+            const current = currentD.getValue()
+            const rels = related && current && related[current]
+            if (rels) {
+              // can this be cached into a derived atom?
+              let best: HoveredWithImportance | undefined
+              for (const rel of Object.values(rels)) {
+                if (!rel) continue
+                if (best && best.importance >= rel.importance) continue
+                best = rel
+              }
+              return best?.relationship
+            }
+            return undefined
+          }
+        })
+      }, [itemKey])
+      return useVal(focusD)
+    },
+    setUserHover(itemKeyOpt) {
+      const prev = currentPresence.get()
+      if (prev === itemKeyOpt) {
+        return
+      }
+      if (prev) {
+        atom.setIn([prev], false)
+      }
+      currentPresence.set(itemKeyOpt)
+      if (itemKeyOpt) {
+        atom.setIn([itemKeyOpt], true)
+      }
+    },
+  }
+}
+
+type RelationshipConfig = {
+  affects: StudioSheetItemKey
+  relationship: Hovered
+  importance?: number
+}
+
+type Presence = {
+  useFocused(itemKey: StudioSheetItemKey | undefined): Hovered | undefined
+  setUserHover(itemKey: StudioSheetItemKey | undefined): void
+  /**  */
+  addRelationships(
+    itemKey: StudioSheetItemKey,
+    config: Array<RelationshipConfig>,
+  ): () => void
+}
+
+const presenceCtx = React.createContext<Presence>(createPresence())
+export function ProvidePresenceRoot({
+  presence = createPresence(),
+  children,
+}: React.PropsWithChildren<{presence?: Presence}>) {
+  return React.createElement(
+    presenceCtx.Provider,
+    {children, value: presence},
+    children,
+  )
+}
+
+const PRESENCE_ITEM_DATA_ATTR = 'data-pi'
```

Comment:
rename `data-pi` -> `data-presence-id`?

Reply:
How about `data-pi-key`? I would like to keep this identifier very short. This identifier is only viewable from within the elements source panel in devtools and nowhere else ATM (future, it might be used to help select in e2e as well)

Reply:
still prefer data-presence-id so we can know its purpose when we see it in devtools. `data-pi-key` isn't so bad, but why is shorter better in this case?